File: serial_dump.py
#!/usr/bin/env python3

from serial import Serial
from struct import unpack
import time

INIT_SEQ: list[list[int]] = [
    [0, 20],  #Set board id to 0
    [135, 0, 100],  #serialdelaytimerwait of 100
    [122, 0, 10],  #number of samples 10
    [123, 0],  #send increment
    [124, 3],  #downsample 3
    [125, 1],  #tickstowait 1
    [136, 2, 32, 0, 0, 255, 200],  # io expanders on (!)
    [136, 2, 32, 1, 0, 255, 200],  # io expanders on (!)
    [136, 2, 33, 0, 0, 255, 200],  #, io expanders on (!)
    [136, 2, 33, 1, 0, 255, 200],  # io expanders on (!)
    [136, 2, 32, 18, 240, 255, 200],  # init
    [136, 2, 32, 19, 15, 255, 200],  # init (and turn on ADCs!)
    [136, 2, 33, 18, 0, 255, 200],  # init
    [136, 2, 33, 19, 0, 255, 200],  # init
    [131, 8, 0],  # adc offset
    [131, 6, 16],  #offset binary output
    # [131, 6, 80], #test pattern output
    [131, 4, 36],  #300 Ohm termination A
    [131, 5, 36],  #300 Ohm termination B
    [131, 1, 0],  #not multiplexed
    [136, 3, 96, 80, 136, 22, 0],  # channel 0 , board 0 calib
    [136, 3, 96, 82, 136, 22, 0],  # channel 1 , board 0 calib
    [136, 3, 96, 84, 136, 22, 0],  # channel 2 , board 0 calib
    [136, 3, 96, 86, 136, 22, 0],  # channel 3 , board 0 calib
]

if __name__ == "__main__":
    ser = Serial("/dev/ttyUSB1", 1500000, timeout=1.0)

    for command in INIT_SEQ:
        ser.write(bytearray(command))
        time.sleep(.1)  # pessimistic init sequence delay

    print("wrote init sequence")
    while True:
        ser.write(bytearray([100, 10]))  # arm trigger and get an event
        result = ser.read(40)
        p = unpack('%dB' % len(result), result)
        for i in range(0, 4):
            print(p[10 * i:10 * i + 10])  # print out the 4 channels * 10 bytes

    ser.close()

File: serial_dump.c
#include <assert.h>
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <unistd.h>

const uint8_t init_seq[] = {
    0,  20,  135, 0,   100, 122, 0,   10,  123, 0,   124, 3,   125, 1,   136,
    2,  32,  0,   0,   255, 200, 136, 2,   32,  1,   0,   255, 200, 136, 2,
    33, 0,   0,   255, 200, 136, 2,   33,  1,   0,   255, 200, 136, 2,   32,
    18, 240, 255, 200, 136, 2,   32,  19,  15,  255, 200, 136, 2,   33,  18,
    0,  255, 200, 136, 2,   33,  19,  0,   255, 200, 131, 8,   0,   131, 6,
    16, 131, 4,   36,  131, 5,   36,  131, 1,   0,   136, 3,   96,  80,  136,
    22, 0,   136, 3,   96,  82,  136, 22,  0,   136, 3,   96,  84,  136, 22,
    0,  136, 3,   96,  86,  136, 22,  0};

const uint8_t sample_seq[] = {100, 10};

int main(int ac, char *av[]) {
  int fd = open("/dev/ttyUSB1", O_SYNC | O_RDWR);

  for (int ii = 0; ii < sizeof(init_seq); ii++) {
    int w = write(fd, &init_seq[ii], 1);
    assert(w == 1);
    usleep(10000); // config settle time
  }
  printf("init sequence wrote\n");

  for (;;) {
    int w = write(fd, sample_seq, sizeof(sample_seq));
    assert(w == 2);

    usleep(10000); // sample acquisition time

    uint8_t read_buf[40];
    int r = read(fd, read_buf, sizeof(read_buf));
    printf("%d\n", r);
    for (int ii = 0; ii < sizeof(read_buf) / 4; ii++) {
      printf("%x ", read_buf[ii]);
    }
    printf("\n");
  }

  return 0;
}
